mantidproject/mantid
# Plot reduced workspace on refl gui preview tab

This was pair-programmed with @ConorMFinn and @thomashampson 

**Description of work.**

This PR plots the result of the preview reduction on the ISIS Reflectometry GUI on a 1D plot. It:
- Adds a new PlotWidget in MantidWidgets which wraps the mplcpp plotting functions and embeds the plot on a QWidget that can be used in C++ GUIs
- Adds an override of the mplcpp `plot` function that takes workspaces by pointer rather than name
- Adds the new widget to the `QtPreviewView`
- Constructs the plot widget presenter in the `PreviewPresenter` and calls into it when the reduction is finished to perform the plot
- Does some minor refactoring in `PreviewPresenter` to ensure all plots are updated on any update
- Does some minor renaming and tidying

Some design notes:
- The axes types are stored in the `QtPlotView` rather than `PlotModel` because the structure required is very Qt- and matplotlib-specific, so logic was required in the view anyway to avoid leaking knowledge of the underlying implementation.
- We have created an interface for `QtPlotView` so that it can be mocked without the mock deriving from a Qt type. We haven't bothered for `PlotModel` because this is a small class only containing standard types. This does mean its methods need to be virtual, but avoids some additional boilerplate that didn't seem worth it.

**To test:**



- Ensure this workspace is loaded or is in your load path: [INTER45455_inst.zip](https://github.com/mantidproject/mantid/files/7504266/INTER45455_inst.zip)
- Open the ISIS Reflectometry GUI
- Select the Preview tab.
- Type in `INTER45455_inst`, set the angle to `1`, and click Load. The instrument view plot should display the data. Click the rectangle-select button above it and select a region. The segments will be summed and the result plotted on the second plot (spectrum vs TOF).
- Click the rectangle-selection button on the second plot and select a region. A reduction will now be performed on the selected spectra and the result plotted on the 1D plot. 
  - You can see the counts better on the second plot if you set the color scale to SymmetricLog10, although there's a bug with autoscaling (un-tick and re-tick to get it to work).
  - Note that the reduction is quite picky and may fail; if so you will get a warning in the log. It's best to select a region around spectra 20-30.
- Click the Export button above the 1D plot. This should export the workspace `preview_reduced_ws` to the ADS. Double click it and you should get the same 1D plot (if you change it to log-log scale).

![image](https://user-images.githubusercontent.com/12895056/177387459-df2a1e43-33e1-4203-8c6e-b844680ee85e.png)


Fixes #34107 

*This does not require release notes* because **it relates to functionality that is not available in the release build yet**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
